test/splitting/annotations.unit

```diff
@@ -0,0 +1,9 @@
+40 columns                              |
+>>> blet
+@veryLongPrefix.VeryLongClassName<VeryLongTypeArgument, OtherVeryLongTypeArgument>(veryLongArgument, otherVeryLongArgument, thirdVeryLongArgument) int i = 0;
+<<<
+@veryLongPrefix.VeryLongClassName<VeryLongTypeArgument, OtherVeryLongTypeArgument>(
```

Comment:
@munificent It seems like this behavior could definitely be improved upon (it's way over 40 columns, and AFACT unnecessarily so).  But I'm not sure how to proceed.  I spent a while trying to modify `visitAnnotation` to make this format better (I was adding calls to `builder.startSpan/endSpan` and `builder.nestExpression/unnest`, trying to mimic what is done in `visitInstanceCreationExpression`).  But I didn't have any luck.  If you want to give me guidance on what else to try I'd be happy to pursue this further.

Reply:
Yeah, this is definitely not ideal. I'll give you some pointers for how to fix this and how to debug issues in general in case you end up doing more hacking on dartfmt. When I want to test its behavior, I use example/format.dart. That formats a given snippet of code and also enables a bunch of diagnostic logging. You can also run that script in a debugger if you want to step through the formatter.

In this case, I changed this line:

```dart
  formatStmt('a is int????;');
```

to:

```dart
  formatUnit('''
@veryLongPrefix.VeryLongClassName<VeryLongTypeArgument, OtherVeryLongTypeArgument>(veryLongArgument, otherVeryLongArgument, thirdVeryLongArgument) int i = 0;
  ''', 40);
```

Now when I run `dart example/format.dart`, it prints:

```
Built:
0:  @veryLongPrefix              ╖      $1  5             nest 0:4
1:  .VeryLongClassName<          ╜╖     $4  TypeArg8      nest 0:4
2:  VeryLongTypeArgument,         ║     $4  TypeArg8      nest 0:4
3:  OtherVeryLongTypeArgument>(   ╜╖    $1  Pos12         nest 0:4
4:  veryLongArgument,              ║2   $1  Pos12         nest 0:4
5:  otherVeryLongArgument,         ║║   $1  Pos12         nest 0:4
6:  thirdVeryLongArgument)         ╜╜       16!                       divide
7:  int                              ╖  $1  18            nest 0:4
8:  i =                              ╜  $1  21            nest 0:4
9:  0;                                      22!
```

This section shows you all of the chunks it creates (roughly its notion of tokens) with the splitting rules attached to the end of each chunk, the spans that group them together (the little bars), and the expression nesting stack in play for each chunk. It's a little (lot) cryptic, but there's a lot of information here.

Then:

```
Writing:
0:  @veryLongPrefix              ╖     $1  5             nest 0:4
1:  .VeryLongClassName<          ╜╖    $4  TypeArg8      nest 0:4
2:  VeryLongTypeArgument,         ║    $4  TypeArg8      nest 0:4
3:  OtherVeryLongTypeArgument>(   ╜╖   $1  Pos12         nest 0:4
4:  veryLongArgument,              ║2  $1  Pos12         nest 0:4
5:  otherVeryLongArgument,         ║║  $1  Pos12         nest 0:4
6:  thirdVeryLongArgument)         ╜╜      16!                       divide
```

This is basically a subset of the above. It prints one of these for each independently splittable region. In this case, it knows the annotation can be handled separately from the variable declaration.

Then it prints the line splitters attempts:

```
? ? ? 1   $0 (106 over) (best)
@veryLongPrefix.VeryLongClassName<VeryLongTypeArgument, OtherVeryLongTypeArgument>(veryLongArgument, otherVeryLongArgument, thirdVeryLongArgument)

1 0 1 1   $4 (10059 over)
@veryLongPrefix
| | .VeryLongClassName<VeryLongTypeArgument, OtherVeryLongTypeArgument>(
| | veryLongArgument, otherVeryLongArgument, thirdVeryLongArgument)

snip many here...

1 2 2 1   $11 (20053 over)
@veryLongPrefix
| | .VeryLongClassName<
| | VeryLongTypeArgument, OtherVeryLongTypeArgument>(veryLongArgument, otherVeryLongArgument,
| | thirdVeryLongArgument)

0 0 4 1   $4 (43 over) (winner)
@veryLongPrefix.VeryLongClassName<VeryLongTypeArgument, OtherVeryLongTypeArgument>(
| | veryLongArgument,
| | otherVeryLongArgument,
| | thirdVeryLongArgument)
```

This shows each attempted solution it came up with and the score it assigned. The leading numbers show which chunks were split, the `$` is the score, and the `(... over)` is how many characters didn't fit within the line length.

Some of those overflow numbers are really large. That's [this hack](https://github.com/dart-lang/dart_style/blob/master/lib/src/line_splitting/solve_state.dart#L382-L404). That gross hack should be rarely hit. Ideally, we'd craft the right set of formatting rules and constraints to completely eliminate it, but I was never fully able to. Argument list formatting is pretty complex.

In this case, you're hitting it because the new type argument list interacts poorly with the following normal argument list. It keeps trying to format things like:

```
@Foo<Bar,
    Baz(argument,
    another)
```

Then it ends up with two lines with the same indentation for different reasons and it punishes itself with a really high score.

Fortunately, normal method calls have a similar problem, so we can just borrow the solution from there. In `visitMethodInvocation()`, we have:

```dart
      builder.nestExpression();
      visit(node.typeArguments);
      visitArgumentList(node.argumentList, nestExpression: false);
      builder.unnest();
```

That outer `nestExpression()` and the `nestExpression: false` bits wrap both the type arguments and value arguments in a single shared level of expression nesting. This way, the first +4 indentation applies to both and then the type argument list will indent another +4 if it splits. Crucially, this avoids the hack where the formatter thinks the two lines have the same indentation for unrelated reasons.

Applied to annotations, it looks like:

```dart
  @override
  void visitAnnotation(Annotation node) {
    token(node.atSign);
    visit(node.name);

    builder.nestExpression();
    visit(node.typeArguments);
    token(node.period);
    visit(node.constructorName);

    // Metadata annotations are always const contexts.
    _constNesting++;

    visitArgumentList(node.arguments, nestExpression: false);

    _constNesting--;

    builder.unnest();
  }
```

Try that and you'll get the much nicer:

```dart
@veryLongPrefix.VeryLongClassName<
        VeryLongTypeArgument,
        OtherVeryLongTypeArgument>(
    veryLongArgument,
    otherVeryLongArgument,
    thirdVeryLongArgument)
int i = 0;
```

Reply:
I think for now I'll just take your suggested change verbatim, but thanks so much for the detailed explanation nonetheless.  It will definitely be useful next time I hack on the formatter :)